File: backend/components/metiers/threed/api/samples/controllers/sample_controllers.js
/* global console */

// Exemple ultra avancé clé en main des contrôleurs API Threed (JS)
// Respecte RGPD, audit, accessibilité, hooks, CI/CD, internationalisation, sécurité, edge cases
const { create, getById, update, delete: del } = require('../controllers/threed_controller');
const { getAuditLog } = require('../audit/audit');
const { rgpdSanitize } = require('../rgpd/rgpd');
const { checkAccessibility } = require('../accessibility/accessibility');
const { beforeAction, afterAction } = require('../hooks/hooks');

async function sampleControllerUltra() {
  // Création avec validation, audit, RGPD, accessibilité
  const data = { name: 'UltraCube', status: 'active', label: 'Ultra', lang: 'fr' };
  beforeAction('create', data);
  const created = await create(data);
  afterAction('create', created);
  checkAccessibility(created);
  const createdSanitized = rgpdSanitize(created);
  console.info('Création:', createdSanitized);

  // Lecture avec audit, RGPD, accessibilité
  const entity = await getById(created.id);
  checkAccessibility(entity);
  const entitySanitized = rgpdSanitize(entity);
  console.info('Lecture:', entitySanitized);

  // Mise à jour
  const updateData = { name: 'UltraCubeV2', status: 'inactive', label: 'UltraV2' };
  beforeAction('update', updateData);
  const updated = await update(created.id, updateData);
  afterAction('update', updated);
  console.info('Mise à jour:', updated);

  // Suppression
  beforeAction('delete', { id: created.id });
  const deleted = await del(created.id);
  afterAction('delete', { id: created.id });
  console.info('Suppression:', deleted);

  // Audit log
  const auditLog = getAuditLog();
  console.log('Audit log:', auditLog);

  // Edge case: accès non autorisé
  try {
    await getById(-1);
  } catch (e) {
    console.warn('Erreur accès:', e);
  }

  console.log('Contrôleur ultra avancé exécuté avec succès.');
}

module.exports = { sampleControllerUltra };

File: backend/components/metiers/threed/api/samples/controllers/sample_controllers.py
# Exemple ultra avancé clé en main des contrôleurs API Threed (Python)
# Respecte RGPD, audit, accessibilité, hooks, CI/CD, internationalisation, sécurité, edge cases
from ..controllers.threed_controller import create_threed, get_threed, update_threed, delete_threed
from ..audit.audit import get_audit_log
from ..rgpd.rgpd import rgpd_sanitize
from ..accessibility.accessibility import check_accessibility
from ..hooks.hooks import before_action, after_action
import logging

def sample_controller_ultra():
    # Création avec validation, audit, RGPD, accessibilité
    data = {'name': 'UltraCube', 'status': 'active', 'label': 'Ultra', 'lang': 'fr'}
    before_action('create', data)
    created = create_threed(data)
    after_action('create', created)
    check_accessibility(created)
    created = rgpd_sanitize(created)
    logging.info(f"Création: {created}")

    # Lecture avec audit, RGPD, accessibilité
    entity = get_threed(created['id'])
    check_accessibility(entity)
    entity = rgpd_sanitize(entity)
    logging.info(f"Lecture: {entity}")

    # Mise à jour
    update_data = {'name': 'UltraCubeV2', 'status': 'inactive', 'label': 'UltraV2'}
    before_action('update', update_data)
    updated = update_threed(created['id'], update_data)
    after_action('update', updated)
    logging.info(f"Mise à jour: {updated}")

    # Suppression
    before_action('delete', {'id': created['id']})
    deleted = delete_threed(created['id'])
    after_action('delete', {'id': created['id']})
    logging.info(f"Suppression: {deleted}")

    # Audit log
    audit_log = get_audit_log()
    print('Audit log:', audit_log)

    # Edge case: accès non autorisé
    try:
        get_threed(-1)
    except Exception as e:
        logging.warning(f"Erreur accès: {e}")

    print('Contrôleur ultra avancé exécuté avec succès.')
